### ISSUE TITLE
It this a complete implementation for Kademlia DHT?

### ISSUE THREAD

**Issue by anticpp** [jech/dht#42]
It's a really good implementation for DHT. 

But I am a little confused with the algorithm for finding nodes.  I think the most wonderful part of 
Kademlia DHT is that we can do a recursive lookup, and each step we can make it `half-more-closer` to the target(aha, I have named it `half-way-iteration-efficent`.). And the most important data struct is how it deals with `buckets`, or `route table`.

As the described in [Ethereum P2P Node Discovery](https://rchain.atlassian.net/wiki/spaces/CORE/pages/15564804/Ethereum+P2P+Node+Discovery+and+Routing).

Simple to say, It make use of an `N` size buckets with the `i` bucket storing the nodes have `i` common prefix as `MyId`.

But, as codes below, It uses a bucket link and iterate the `3` nearby buckets and pick the `k` nearest nodes. I have not figured out how It can make a `half-way-iteration-efficent`? Or if there something wrong understand to me, please sort me out, will be thankful.

```
b = find_bucket(id, AF_INET);
if(b) {
    numnodes = buffer_closest_nodes(nodes, numnodes, id, b);
     if(b->next)
         numnodes = buffer_closest_nodes(nodes, numnodes, id, b->next);
     b = previous_bucket(b);
     if(b)
        numnodes = buffer_closest_nodes(nodes, numnodes, id, b);
}
```

**Comment by jech** [comment]
You're misundestanding.  The code you've quoted buffers the nodes closest to the *target* that are known to the current peer.  In a fully converged DHT, these are at least halfway to the target.

The implementation follows fairly closely the algorithm of the BitTorrent DHT, with a few tweaks for increased convergence (thanks, @ghazel).  It has some limitations wrt. the algorithm described in the Kademlia paper, but that's common to all implementations of the BitTorrent DHT.

**Comment by anticpp** [comment]
Well, thanks for your response. You are right.

I have reviewed the Kademlia paper couple of times, after reading your implementation. 
However, I found that I have misunderstood the `bucket split` part, or saying, I missed some details.
And this is the really important part of the Kademlia bucket.

Be more concrete, each node only splits the bucket of his `my bucket`. That means, the more splitted bucket maintains the nearer distance by `my node`. 

And last, the algorithm of Kademlia is still too obscure to be understood.(*-*)
